src/mon/MonClient.cc

```diff
@@ -687,23 +688,49 @@ MonConnection& MonClient::_add_conn(unsigned rank, uint64_t global_id)
 
 void MonClient::_add_conns(uint64_t global_id)
 {
-  uint16_t min_priority = std::numeric_limits<uint16_t>::max();
+  map<uint16_t, vector<unsigned>> rank_by_priority;
   for (const auto& m : monmap.mon_info) {
-    if (m.second.priority < min_priority) {
-      min_priority = m.second.priority;
-    }
+    rank_by_priority[m.second.priority].push_back(monmap.get_rank(m.first));
+  }
+  vector<unsigned> candidates;
+  if (!rank_by_priority.empty()) {
+    // pick mon by priority, lowest first
+    // see http://tracker.ceph.com/issues/5249 for the use case of priority
+    candidates = rank_by_priority.begin()->second;
   }
   vector<unsigned> ranks;
-  for (const auto& m : monmap.mon_info) {
-    if (m.second.priority == min_priority) {
-      ranks.push_back(monmap.get_rank(m.first));
+  if (candidates.size() > 1) {
+    // we got multiple candidates
+    // see if we need to make our selection based on their corresponding weight
+    vector<int> weights;
+    for (auto i : candidates) {
+      auto rank_name = monmap.get_name(i);
+      auto rank_weight = monmap.get_weight(rank_name);
+      // filter out any candidate without a valid (non-zero) weight
+      if (rank_weight) {
+        ranks.push_back(i);
+        weights.push_back(rank_weight);
+      }
+    }
+    if (ranks.empty()) {
+      // good, no weights specified
+      // shuffle so all targets get equal attention
+      std::swap(ranks, candidates);
+      std::random_device rd;
+      std::mt19937 gen(rd());
+      std::shuffle(ranks.begin(), ranks.end(), gen);
+    } else if (ranks.size() == 1) {
+      // fast out for single target
+    } else {
+      ceph_assert(ranks.size() == weights.size());
+      std::random_device rd;
+      std::mt19937 gen(rd());
+      weighted_shuffle(ranks, weights, gen);
     }
   }
-  std::random_device rd;
-  std::mt19937 rng(rd());
-  std::shuffle(ranks.begin(), ranks.end(), rng);
+  ldout(cct, 10) << __func__ << " ranks=" << ranks << dendl;
   unsigned n = cct->_conf->mon_client_hunt_parallel;
-  if (n == 0 || n > ranks.size()) {
+  if (n == 0 || n > ranks.size()) { // always choose first N
```

Comment:
this comment is confusing. we are not choosing the "first N" here. we are connecting to *all* of the monitor candidates.